src/engine/sched.c

```diff
@@ -1150,67 +1299,119 @@ should_enqueue_req(struct dss_xstream *dx, struct sched_req_attr *attr)
 	return dx->dx_main_xs;
 }
 
-static void
+static int
 req_enqueue(struct dss_xstream *dx, struct sched_request *req)
 {
 	struct sched_info	*info = &dx->dx_sched_info;
 	struct sched_pool_info	*spi = req->sr_pool_info;
 	struct sched_req_attr	*attr = &req->sr_attr;
 	struct sched_req_info	*sri;
+	int			 rc = 0;
 
 	D_ASSERT(spi != NULL);
+	D_ASSERT(attr->sra_type < SCHED_REQ_MAX);
 	sri = &spi->spi_req_array[attr->sra_type];
 
+	D_ASSERT(req->sr_in_heap == 0);
 	D_ASSERT(d_list_empty(&req->sr_link));
 	if (attr->sra_type == SCHED_REQ_UPDATE ||
 	    attr->sra_type == SCHED_REQ_FETCH) {
 		D_ASSERT(policy_ops[sched_policy].enqueue_io != NULL);
-		policy_ops[sched_policy].enqueue_io(dx, req, NULL);
+		rc = policy_ops[sched_policy].enqueue_io(dx, req, NULL);
 	} else {
 		d_list_add_tail(&req->sr_link, &sri->sri_req_list);
 	}
 	req->sr_enqueue_ts = info->si_cur_ts;
 
 	sri->sri_req_cnt++;
 	spi->spi_req_cnt++;
-	info->si_req_cnt++;
+	info->si_total_req_cnt++;
+	info->si_req_cnt[attr->sra_type]++;
+
+	return rc;
+}
+
+#define MAX_SCHED_REQ_NUM	(2 << 20)
+
+static bool
+req_need_reject(struct sched_req_attr *attr, struct sched_info *info)
+{
+	uint64_t	estimated_time = 0;
+	uint64_t	req_num = 0;
+	int		i;
+
+	/* Old clients RPC won't be rejected */
+	if (attr->sra_flags & SCHED_REQ_FL_NO_REJECT)
+		return false;
+
+	/*
+	 * Calculate time based on ults on argobots and non-system
+	 * requests queued. It is not easy to estimate how many system
+	 * ults will be executed, reserve 50% of timeout for it.
+	 */
+	for (i = SCHED_REQ_UPDATE; i < SCHED_REQ_MAX; i++) {
+		req_num += info->si_kicked_req_cnt[i];
+		if (!is_system_req(i)) {
+			estimated_time += info->si_req_cnt[i] * req_latencys[i];
+			req_num += info->si_req_cnt[i];
+		}
+	}
+	/* convert to msecs */
+	estimated_time /= 1000;
+	estimated_time += MAX_CYCLE_TIME;
+	if ((estimated_time << 1) > attr->sra_timeout)
```

Comment:
Isn't it too aggressive to reject an RPC when the (estimated time) > (timeout / 2)?